### PR TITLE
Implement anonymous uploads, fixes #4250

### LINKED ISSUE #4250: [S3 gateway] Anonymous upload of object in a bucket with "upload" policy fails with Access Denied
## Expected Behavior

Anonymous upload should succeed on a bucket with "upload" policy.

## Current Behavior

Anonymous upload fails with Access Denied on a bucket with "upload" policy.
```
➜  minio git:(master) ✗ mc policy mys3/test/prefix/obj1
Access permission for `mys3/test/prefix/obj1` is `upload`
```
```
➜  minio git:(master) ✗ curl --upload-file /etc/lsb-release localhost:9001/test/prefix/obj1 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0   105    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0<?xml version="1.0" encoding="UTF-8"?>
100   341    0   236  100   105    234    104  0:00:01  0:00:01 --:--:--   234
<Error><Code>AccessDenied</Code><Message>Access Denied.</Message><Key></Key><BucketName></BucketName><Resource>/test/prefix/obj1</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>%                               
```

## Possible Solution

Not known yet
## Steps to Reproduce (for bugs)

1. `mc policy upload mys3/testbucket/prefix`
2. `curl --upload-file /etc/lsb-release localhost:9001/testbucket/prefix/s1`, where gateway is running on `localhost:9001`

## Your Environment

* Version used (`minio version`): 
```
Version: 2017-05-04T11:56:54Z
Release-Tag: DEVELOPMENT.2017-05-04T11-56-54Z
Commit-ID: 0ea8bfaf78ba633b5951fa1d9c604b3b50a4dd62

```
* Environment name and version (e.g. nginx 1.9.1):
* Server type and version:
* Operating System and version (`uname -a`):
* Link to your project:

### PR DESCRIPTION
**Pull request by nl5887**

### CODE REVIEW

File: cmd/gateway-handlers.go [hunk 1]
```diff
@@ -151,6 +154,124 @@ func (api gatewayAPIHandlers) GetObjectHandler(w http.ResponseWriter, r *http.Re
...
+	w.Header().Set("ETag", "\""+objInfo.MD5Sum+"\"")
+	writeSuccessResponseHeadersOnly(w)
+
+	// Get host and port from Request.RemoteAddr.
+	host, port, err := net.SplitHostPort(r.RemoteAddr)
+	if err != nil {
+		host, port = "", ""
+	}
+
+	// Notify object created event.
```
Comment (harshavardhana) [inline cmd/gateway-handlers.go]: We don't need this part of the code for Gateway. since we won't implement notifications in Gateway.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/gateway-azure-anonymous.go [hunk 2]
```diff
@@ -57,6 +57,28 @@ func (a AzureObjects) AnonGetBucketInfo(bucket string) (bucketInfo BucketInfo, e
...
+		return ObjectInfo{}, azureToObjectError(traceError(err), bucket, object)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return ObjectInfo{}, azureToObjectError(traceError(anonErrToObjectErr(resp.StatusCode, bucket, object)), bucket, object)
+	}
+
+	return a.AnonGetObjectInfo(bucket, object)
+}
```
Comment (krishnasrinivas) [inline cmd/gateway-azure-anonymous.go]: Azure does not support anonymous PUT, so here you can just return NotImplemented

File: cmd/gateway-s3-anonymous.go [hunk 3]
```diff
@@ -17,11 +17,42 @@
...
+	md5sum := metadata["md5Sum"]
+	if md5sum != "" {
+		md5sumBytes, err = hex.DecodeString(md5sum)
+		if err != nil {
+			return ObjectInfo{}, s3ToObjectError(traceError(err), bucket, object)
+		}
+		delete(metadata, "md5Sum")
+	}
+
+	oi, err := l.Client.PutObject(bucket, object, size, data, md5sumBytes, sha256sumBytes, toMinioClientMetadata(metadata))
```
Comment (krishnasrinivas) [inline cmd/gateway-s3-anonymous.go]: l.anonClient should be used

File: cmd/gateway-handlers.go [hunk 4]
```diff
@@ -151,6 +154,124 @@ func (api gatewayAPIHandlers) GetObjectHandler(w http.ResponseWriter, r *http.Re
...
+	bucket = vars["bucket"]
+	object = vars["object"]
+
+	objectAPI := api.ObjectAPI()
+	if objectAPI == nil {
+		writeErrorResponse(w, ErrServerNotInitialized, r.URL)
+		return
+	}
+
+	reqAuthType := getRequestAuthType(r)
```
Comment (krishnasrinivas) [inline cmd/gateway-handlers.go]: this is repeated again at line 190

File: cmd/gateway-handlers.go [hunk 5]
```diff
@@ -151,6 +154,124 @@ func (api gatewayAPIHandlers) GetObjectHandler(w http.ResponseWriter, r *http.Re
...
+			errorIf(err, "Unable to parse `x-amz-decoded-content-length` into its integer value", sizeStr)
+			writeErrorResponse(w, toAPIErrorCode(err), r.URL)
+			return
+		}
+	}
+	if size == -1 {
+		writeErrorResponse(w, ErrMissingContentLength, r.URL)
+		return
+	}
+
```
Comment (krishnasrinivas) [inline cmd/gateway-handlers.go:201]: signature is not being verified when authtype is authTypeSignedV2, authTypePresignedV2, authTypePresigned, authTypeSigned

**Review by krishnasrinivas — CHANGES_REQUESTED** [review]

**Review by krishnasrinivas — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**